src/Symfony/Component/Cache/Adapter/AbstractAdapter.php

```diff
@@ -259,13 +248,22 @@ public function commit()
             if (false === $ok) {
                 $ok = array_keys($values);
             }
-            foreach ($ok as $failedId) {
-                $key = substr($failedId, $namespaceLen);
-                $ko[$key] = $this->deferred[$key];
+            foreach ($ok as $id) {
+                $ko[$lifetime][] = array($id => $values[$id]);
+            }
+        }
+
+        $this->deferred = array();
+        $ok = true;
+
+        // When bulk-save failed, retry each item individually
```

Comment:
The previous logic has a "memory leak" in that whenever a deferred couldn't saved, it would never be freed.
The logic I propose here is: try bulk-save, then try individual save, then trash.